File: 0501~1000/0962/Maximum Width Ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        stack<int> idx;
        int n = nums.size();
        for(int i = 0;i < n;i ++){
            if(idx.empty() || nums[i] < nums[idx.top()])
                idx.push(i);
        }

        int ans = 0;
        for(int i = n - 1;i >= 0;i --){
            while(!idx.empty() && nums[idx.top()] <= nums[i]){
                ans = max(ans, i - idx.top());
                idx.pop();
            }
        }
        return ans;
    }
};

File: 0501~1000/0962/Maximum Width Ramp.py
class Solution:
    def maxWidthRamp(self, nums: List[int]) -> int:
        stack = []
        n = len(nums)
        sz = 0
        for i in range(n):
            if sz == 0 or nums[stack[sz - 1]] > nums[i]:
                stack.append(i)
                sz += 1
        
        ans = 0
        for i in range(n - 1, 0, -1):
            while sz > 0 and nums[stack[sz - 1]] <= nums[i]:
                ans = max(ans, i - stack[sz - 1])
                stack.pop()
                sz -= 1
        return ans
